_posts/2017-10-26-totw-94.md

```diff
@@ -0,0 +1,138 @@
+---
+title: "Tip of the Week #94: Callsite Readability and bool Parameters"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/94
+type: markdown
+order: 094
+---
+
+Originally posted as totw/94 on 2015-04-27
+
+*By Geoff Romer [(gromer@google.com)](mailto:gromer@google.com)*
+
+Revised 2017-10-25
+
+*"Of the many forms of false culture, a premature converse with abstractions is
+perhaps the most likely to prove fatal to the growth of &hellip; vigour of
+intellect." &mdash; George Boole.*
+
+Pop quiz: what does this code do?
+
+```c++ {.bad}
+int main(int argc, char* argv[]) {
+  ParseCommandLineFlags(&argc, &argv, false);
+}
+```
+
+In particular, can you explain what the last argument means without looking at
+the documentation? I'll give you a hint: it has to do with whether google3
+command-line flags are left in `argv` after the call, but which is `true` and
+which is `false`?
+
+OK, time's up. Pencils down. The answer is that this call will leave `argv`
+untouched. As it happens, `ParseCommandLineFlags` removes command-line flags
+from `argv` if the third argument is true.
+
+If you didn't know that, good. You have better things to do with your brain than
+memorize the parameter lists of semi-obscure google3 functions. But here's the
```

Comment:
Another mention of google3. Also, this whole passage assumes that the reader is vaguely familiar with `ParseCommandLineFlags()`, or at least could be. It doesn't really work if it's a made up function.

Our best option is probably to restructure this passage as discussing a hypothetical example. Something like:

Suppose you come across code like this:

```c++ {.bad}
￼int main(int argc, char* argv[]) {
￼  ParseCommandLineFlags(&argc, &argv, false);
￼}
```

Can you tell what this code does, and in particular what the last argument means? Now suppose you've come across this function before, and you know that the last argument has to do with whether command-line flags are left in `argv` after the call. Can you tell which is `true` and which is `false`?

Of course you don't know, because this is hypothetical, but even in real code, we have better things to do with their brains than memorize the meaning of every function parameter, and better things to do with our time than go look up the documentation for every function call we encounter. We have to be able to make pretty good guesses about what function calls mean, just from looking at the callsite.

Reply:
I like it.